### PR TITLE
Enable rocm

### PR DESCRIPTION
**Pull request by micmelesse**
This PR enables trition on rocm. There is still more work to do but we are able to compile on rocm and run an empty kernel. This is the first of many prs. 

This pr uses a lot of ifdefs to work around some of the limitations of the current version of hipify torch. We can remove them as hipify torch is updated.  

That being said, in order for this PR to work, this PR in hipify torch must be merged https://github.com/ROCmSoftwarePlatform/hipify-torch/pull/7. It basically removes bad unicode characters from source code

The major changes are in the following files
- include/triton/driver/dispatch.h
- lib/driver/dispatch.cc
- lib/driver/module.cc
- CMakeLists.txt

### CODE REVIEW

File: lib/driver/module.cc [hunk 1]
```diff
@@ -252,12 +286,21 @@ std::string cu_module::compile_llvm_module(llvm::Module* module, driver::device*
   int ptx = vptx(version);
   int ptx_major = ptx / 10;
   int ptx_minor = ptx % 10;
+#endif
   // create
   llvm::SmallVector<char, 0> buffer;
+#ifdef __HIP_PLATFORM_AMD__
+  std::string triple = "amdgcn-amd-amdhsa";
+  std::string proc = "gfx908"; //TODO: add code to QUERY GPU to make sure it works
```
Comment (jithunnair-amd) [inline lib/driver/module.cc:294]: This hardcoding will need to be fixed in future PRs. As discussed with @micmelesse earlier, if we don't have a HIP API to provide us the gfx arch, we will need to request that from HIP team in a JIRA. cc @sunway513

File: python/setup.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def build_extension(self, ext):
         if not os.path.exists(llvm_build_dir):
             os.makedirs(llvm_build_dir)
         # python directories
-        python_include_dirs = [distutils.sysconfig.get_python_inc()] + ['/usr/local/cuda/include']
+        python_include_dirs = [distutils.sysconfig.get_python_inc()] + ['/usr/local/cuda/include'] + ['/opt/rocm/include']
```
Comment (jithunnair-amd) [inline python/setup.py:91]: This will likely need to be conditionalized on ROCm/CUDA using TRITON_USE_ROCM env variable.

**Review by jithunnair-amd — APPROVED** [review]
Approving as a proof-of-concept PR. A lot of the ifdefs will be rendered redundant once hipify-torch is updated to hipify the header paths. Some changes will need to be made to generalize the ROCm path (eg. "gfx908"). I cannot comment on the LLVM logic as that's not my area of expertise.

File: lib/driver/module.cc (hunk 1 above)
Comment (micmelesse) [reply lib/driver/module.cc]: I add a function to grep for this info using rocminfo.

File: CMakeLists.txt [hunk 3]
```diff
@@ -24,29 +32,37 @@ find_library(TERMINFO_LIBRARY tinfo)
 
 # Compiler flags
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D__STDC_FORMAT_MACROS  -std=gnu++17")
+if (TRITON_USE_ROCM)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D__STDC_FORMAT_MACROS -fvisibility=default -std=gnu++17 -Wno-unused-result -Wno-attributes")
```
[the remark this answers is not in the document]
Comment (micmelesse) [reply CMakeLists.txt:36]: Some of them where to suppress warnings that occurred due to the hipified code. I will get rid of the  -fvisibility=default

File: lib/codegen/pass.cc [hunk 4]
```diff
@@ -34,7 +40,11 @@ void add_passes_to_emit_bin(ir::module &ir, driver::device *dev, int num_warps,
   std::unique_ptr<llvm::Module> llvm(new llvm::Module(name, ctx));
   // optimizations
   std::unique_ptr<codegen::target> target = dev->make_target();
+#ifdef __HIP_PLATFORM_AMD__
+  bool cts_use_async = false;
```
[the remark this answers is not in the document]
Comment (micmelesse) [reply lib/codegen/pass.cc:44]: I just need to pick a value for this variable it was causing segfaults on AMD

File: python/setup.py (hunk 2 above)
Comment (micmelesse) [reply python/setup.py]: Done

**Comment by micmelesse** [comment]
This got merged upstream https://github.com/openai/triton/pull/231. So I am going to merge it.

**Pull request merged**